### ISSUE TITLE
Invalid Cursor State when using executemany

### ISSUE THREAD

**Issue by boumboum** [mkleehammer/pyodbc#321]
When I try to insert loads of data with some columns (10-15) using executemany (and fast_executemany = True) I receive Invalid Cursor State (0) (SQLPrepare) error.
I suspect the params array being too big either for the driver, or for SQL Server or for pyodbc.
Is there any other way to "emulate" bcp?
Or is there any workaround/setting?

Environment:
- Python: 3.6.3 (64 bit)
- pyodbc: 4.0.21
- OS: Win7 (64 bit)
- DB: MSSQL 2014
- driver: ODBC 13

Edit:
After googling it, it seams, there is a limit of 1000 of the "values" clause.
is pyodbc using for executemany this statement: instert into mytable (c1, c2, ...) values (v1, v2, ...), (v1, v2, ...) ? even if fast_executemany is set to True?
How to overcome of this issue?
  
  Edit2:
It turned out, that the problem doesn't come from a limit issue.
After testing several use cases I've realized that if there is any cr.execute BEFORE cr.executemany than the cursor gets invalidated. 
One workaround is to use a different cursor for executemany or (re)declare the cursor before each usage (which is not too elegant)

**Comment by v-chojas** [comment]
I believe I have found the cause, and fixed it in this PR: https://github.com/mkleehammer/pyodbc/pull/327

You can try the fix here: https://github.com/v-chojas/pyodbc/tree/fix-fem-curstate

**Comment by pinkerltm** [comment]
I am experiencing the same issue (pyodbc 4.0.21). Currently I am working on a Windows box without C++ dev Environment, so it will take some time to compile and test the fix.

**Comment by pinkerltm** [comment]
@v-chojas AFter compiling your fix and running my app with pyodbc 4.0.19b I am still getting the same error message. It first INSERTS 40.000 rows, then commits, and then tries to insert another 40.000 rows. At this Point, I get "Invalid Cursor state"

Python 3.5.2 Win 10 64 bit

**Comment by v-chojas** [comment]
@pinkerltm please make sure you are using the fixed pyodbc. This works for me with the fix, but gives the invalid cursor state without:

```
import pyodbc
conn = pyodbc.connect('YOUR CONNECTION STRING HERE')
stmt = conn.cursor()
stmt.fast_executemany = True
stmt.execute('SELECT 1')
stmt.executemany('INSERT INTO TestTable(c1) values(?)', [[1],[2],[3],[4],[5]])
stmt.commit()
```

**Comment by boumboum** [comment]
@v-chojas : Unfortunately I cannot compile your code on my windows box (so I will wait until I can install it with pip and I will use my workaround (a second cursor)) but if I understood pinkerltm well, could you modify your code inserting also commits in between to see what happens like that:
```
import pyodbc
conn = pyodbc.connect('YOUR CONNECTION STRING HERE')
stmt = conn.cursor()
stmt.fast_executemany = True
stmt.execute('SELECT 1')
stmt.executemany('INSERT INTO TestTable(c1) values(?)', [[1],[2],[3],[4],[5]])
stmt.commit()
res = stmt.execute('SELECT * from TestTable').fetchall()
stmt.executemany('INSERT INTO TestTable(c1) values(?)', [[1],[2],[3],[4],[5]])
stmt.commit()
```

**Comment by v-chojas** [comment]
With my fix, that works too, and without, yields the same ICS error.

**Comment by boumboum** [comment]
Great! :-)
In this case I'm looking forward to get integrated it ASAP into the official release to be able to install via pip. (my customers do not like "unofficial" builds. :-( )

@Michael: It's up to you :-)

**Comment by pinkerltm** [comment]
@v-chojas sorry mate, I tried the following:

1. pip uninstall pyodbc (removed pyodbc 4.0.21)
1. Cloned your repo ... python Setup.py build .... python Setup.py install (pip freeze tells me now pyodbc==4.0.18b12
1. Created a new python file pyodbc_fastem_testcode.py with exactly your Code (only a table which exists in my database)
1. run it (again that gives me "Invalid Cursor state")

Do you know another way how I can be sure, that the fixed pyodbc is used? Any other suggestions what I might do wrong?

**Comment by v-chojas** [comment]
Try this from interactive prompt:
```
import pyodbc, os, time
time.ctime(os.path.getmtime(pyodbc.__file__))
```
If it's the one you built and installed, its date should correspond.

**Comment by mkleehammer** [comment]
Released in 4.0.22

**Comment by pinkerltm** [comment]
@v-chojas JFYI: I could not sort out why the Code worked not for me, after all I upgraded back to the Version @mkleehammer had released recently and now it works perfect. Many thanks for speeding up my ETL Tasks, i.e. INSERTING 40 Mio. rows, by a factor of 10 (!). Thats fantastic work!
